File: packages/sdk-typescript/src/scenes/http.ts
import { randomUUID } from "node:crypto";
import { gunzipSync, inflateSync, brotliDecompressSync } from "node:zlib";
import type { Playback } from "./playback.js";
import { CaptureSession } from "./capture.js";
import { sceneContext, type SceneRuntime } from "./context.js";
import { encodePayload } from "./payload.js";
import {
  ARTIFACT_BYTES,
  canonical,
  cleanHeaders,
  cleanUrl,
  sanitize,
  sha256,
  SnapshotMissError,
} from "./portable.js";
import type { Binding, HttpPayload } from "./types.js";
export const SEMANTIC_HEADERS = [
  "accept",
  "content-type",
  "range",
  "if-match",
  "if-none-match",
  "if-modified-since",
  "if-unmodified-since",
  "if-range",
];
function isJsonContentType(value: string | null | undefined): boolean {
  const mediaType = value?.split(";")[0].trim().toLowerCase() ?? "";
  return (
    mediaType === "application/json" ||
    /^[!#$%&'*+.^_`|~0-9a-z-]+\/[!#$%&'*+.^_`|~0-9a-z-]+\+json$/.test(mediaType)
  );
}
export function httpBinding(
  runtime: SceneRuntime,
  url: string,
): Binding | undefined {
  const u = new URL(url);
  const bindings = runtime.bindings.filter(
    (b) =>
      runtime.selected(b.id) &&
      b.kind === "http" &&
      b.http?.origin === u.origin &&
      u.pathname.startsWith(b.http.pathPrefix),
  );
  if (bindings.length > 1) throw new SnapshotMissError("overlapping_bindings");
  return bindings[0];
}
export function httpArguments(
  binding: Binding,
  request: Request,
  bytes: Uint8Array,
) {
  let body = bytes;
  if (isJsonContentType(request.headers.get("content-type")) && bytes.length) {
    const decoded = new TextDecoder("utf-8", { fatal: true }).decode(bytes);
    body = Buffer.from(canonical(sanitize(JSON.parse(decoded)).value));
  }
  return {
    method: request.method.toUpperCase(),
    url: cleanUrl(request.url),
    headers: cleanHeaders(request.headers, [
      ...SEMANTIC_HEADERS,
      ...(binding.http?.headers ?? []).map((h) => h.toLowerCase()),
    ]),
    bodySha256: sha256(body),
  };
}
async function fetchBody(
  input: RequestInfo | URL,
  init?: RequestInit,
): Promise<Uint8Array> {
  const b = init?.body;
  if (b === undefined || b === null) {
    if (input instanceof Request && input.body)
      throw new TypeError(
        "Streaming Request bodies require an enclosing source tool",
      );
    return new Uint8Array();
  }
  if (typeof b === "string" || b instanceof URLSearchParams)
    return Buffer.from(String(b));
  if (b instanceof ArrayBuffer) return new Uint8Array(b);
  if (ArrayBuffer.isView(b))
    return new Uint8Array(b.buffer, b.byteOffset, b.byteLength);
  if (b instanceof Blob && b.size <= ARTIFACT_BYTES)
    return new Uint8Array(await b.arrayBuffer());
  throw new TypeError("Streaming and multipart requests are not portable");
}
type CaptureHandle = ReturnType<CaptureSession["begin"]>;
/** Bounded accumulator; capturing never drains a response that the application has not consumed. */
export class HttpCapture {
  private chunks: Uint8Array[] = [];
  private size = 0;
  private eligible = true;
  private ended = false;
  constructor(
    readonly capture: CaptureSession,
    readonly handle: CaptureHandle,
    readonly metadata: Omit<HttpPayload, "body">,
  ) {}
  chunk(value: Uint8Array) {
    if (!this.eligible || this.ended) return;
    try {
      if (
        this.size + value.byteLength > ARTIFACT_BYTES ||
        !this.capture.client.reserve(value.byteLength)
      )
        throw new Error();
      this.size += value.byteLength;
      this.chunks.push(new Uint8Array(value));
    } catch {
      this.disable();
    }
  }
  private disable() {
    this.eligible = false;
    this.capture.client.release(this.size);
    this.size = 0;
    this.chunks = [];
  }
  finish() {
    if (this.ended) return;
    this.ended = true;
    try {
      const bytes = new Uint8Array(Buffer.concat(this.chunks));
      if (
        this.eligible &&
        bytes.length > 0 &&
        isJsonContentType(this.metadata.headers["content-type"])
      ) {
        try {
          if (
            sanitize(
              JSON.parse(
                new TextDecoder("utf-8", { fatal: true }).decode(
                  decompress(bytes, this.metadata.headers["content-encoding"]),
                ),
              ),
            ).changed
          )
            this.eligible = false;
        } catch {
          this.eligible = false;
        }
      }
      this.capture.client.release(this.size);
      this.size = 0;
      this.chunks = [];
      this.capture.finishPayload(
        this.handle,
        async () => {
          let body;
          const mime =
            this.metadata.headers["content-type"]?.split(";")[0].trim() ??
            "application/octet-stream";
          if (
            this.eligible &&
            bytes.length > 0 &&
            this.metadata.status >= 200 &&
            this.metadata.status < 300 &&
            (mime === "application/pdf" ||
              mime ===
                "application/vnd.openxmlformats-officedocument.wordprocessingml.document" ||
              mime ===
                "application/vnd.openxmlformats-officedocument.presentationml.presentation" ||
              this.metadata.headers["content-disposition"]
                ?.toLowerCase()
                .includes("attachment"))
          ) {
            const sourceBytes = decompress(
              bytes,
              this.metadata.headers["content-encoding"],
            );
            const ref = await this.capture.client.upload(
              sourceBytes,
              mime,
              "source",
              "source",
            );
            void this.capture.source({
              id: randomUUID(),
              relation: "tool_source",
              reference: ref,
              callId: this.handle?.callId,
              mimeType: mime,
              ...(this.metadata.url ? { uri: this.metadata.url } : {}),
              content: this.metadata.status === 206 ? "partial" : "complete",
            });
            body = this.metadata.headers["content-encoding"]
              ? await encodePayload(this.capture.client, bytes)
              : { kind: "blob" as const, ref };
          } else body = await encodePayload(this.capture.client, bytes);
          return { ...this.metadata, body };
        },
        bytes.length,
        this.eligible,
        this.eligible ? undefined : "http_body_cap_or_redacted",
      );
    } catch {
      this.disable();
      this.capture.finishPayload(
        this.handle,
        async () => ({ kind: "absent" }),
        0,
        false,
        "unsupported_http_response",
      );
    }
  }
  fail(error: unknown, cancelled = false) {
    if (this.ended) return;
    this.ended = true;
    this.disable();
    // The application has already received an HTTP response. A function-level error
    // cannot reproduce its headers or partially consumed body during playback.
    if (this.handle) this.handle.replayable = false;
    this.capture.fail(this.handle, error, cancelled);
  }
}
export function responseMetadata(
  response: Response,
  decoded = true,
): Omit<HttpPayload, "body"> {
  const headers = cleanHeaders(response.headers);
  delete headers["transfer-encoding"];
  if (
    decoded &&
    response.body &&
    headers["content-encoding"] &&
    headers["content-encoding"] !== "identity"
  ) {
    delete headers["content-encoding"];
    delete headers["content-length"];
  }
  return {
    kind: "http",
    status: response.status,
    statusText: response.statusText,
    headers,
    ...(response.url ? { url: cleanUrl(response.url) } : {}),
  };
}
function decompress(bytes: Uint8Array, encoding?: string): Uint8Array {
  switch (encoding?.toLowerCase()) {
    case undefined:
    case "identity":
      return bytes;
    case "gzip":
      return gunzipSync(bytes, { maxOutputLength: ARTIFACT_BYTES });
    case "deflate":
      return inflateSync(bytes, { maxOutputLength: ARTIFACT_BYTES });
    case "br":
      return brotliDecompressSync(bytes, { maxOutputLength: ARTIFACT_BYTES });
    default:
      throw new SnapshotMissError("nonportable");
  }
}
export function responseFromRecorded(
  value: unknown,
  decoded = false,
  head = false,
): Response {
  if (
    !value ||
    typeof value !== "object" ||
    !("kind" in value) ||
    value.kind !== "http" ||
    !("body" in value) ||
    !(value.body instanceof Uint8Array)
  )
    throw new SnapshotMissError("incompatible");
  const p = value as unknown as Omit<HttpPayload, "body"> & {
    body: Uint8Array;
  };
  if (p.status < 200 || p.status > 599)
    throw new SnapshotMissError("nonportable");
  const headers = { ...p.headers };
  const transformed =
    decoded &&
    !head &&
    p.body.length > 0 &&
    Boolean(headers["content-encoding"]) &&
    headers["content-encoding"] !== "identity";
  const body = transformed
    ? decompress(p.body, headers["content-encoding"])
    : p.body;
  if (transformed) {
    delete headers["content-encoding"];
    delete headers["content-length"];
  }
  const response = new Response(
    head || [204, 205, 304].includes(p.status) ? null : new Uint8Array(body),
    { status: p.status, statusText: p.statusText, headers },
  );
  if (p.url) Object.defineProperty(response, "url", { value: p.url });
  return response;
}
export function wrapFetch(
  fetcher: typeof fetch = globalThis.fetch.bind(globalThis),
): typeof fetch {
  const wrapped: typeof fetch = (async (
    input: RequestInfo | URL,
    init?: RequestInit,
  ) => {
    const active = sceneContext.getStore();
    if (!active || active.suppressed) return fetcher(input, init);
    const request =
      input instanceof Request
        ? new Request(input.url, {
            method: init?.method ?? input.method,
            headers: init?.headers ?? input.headers,
            signal: init?.signal ?? input.signal,
            ...(init?.body ? { body: init.body } : {}),
          })
        : new Request(input, init);
    const binding = httpBinding(active.runtime, request.url);
    if (!binding || !active.runtime.selected(binding.id))
      return fetcher(input, init);
    let args: unknown;
    let eligible = true;
    try {
      if (
        request.headers
          .get("content-type")
          ?.toLowerCase()
          .startsWith("multipart/")
      )
        throw new TypeError();
      const bytes = await fetchBody(input, init);
      if (bytes.length > ARTIFACT_BYTES) throw new TypeError();
      args = httpArguments(binding, request, bytes);
    } catch {
      eligible = false;
      args = {
        method: request.method,
        url: cleanUrl(request.url),
        unsupportedBody: true,
      };
    }
    if (active.runtime.mode === "playback") {
      if (!eligible)
        (active.runtime as Playback).reject(
          binding.id,
          "http",
          args,
          "nonportable",
        );
      return responseFromRecorded(
        active.runtime.invoke(binding.id, "http", args, () => {
          throw new SnapshotMissError("unrecorded");
        }),
        true,
        request.method === "HEAD",
      );
    }
    const capture = active.runtime as CaptureSession;
    const handle = capture.begin(binding.id, "http", args);
    if (!eligible && handle) handle.replayable = false;
    let response: Response;
    try {
      response = await sceneContext.run(
        { ...active, parentCallId: handle?.callId, suppressed: true },
        () => fetcher(input, init),
      );
    } catch (e) {
      capture.fail(handle, e, request.signal.aborted);
      throw e;
    }
    try {
      const observer = new HttpCapture(
        capture,
        handle,
        responseMetadata(response),
      );
      if (!response.body) {
        observer.finish();
        return response;
      }
      if (response.status < 200 || response.status > 599) {
        observer.fail(new Error("Unsupported status"), true);
        return response;
      }
      const reader = response.body.getReader();
      const stream = new ReadableStream<Uint8Array>(
        {
          async pull(controller) {
            try {
              const part = await reader.read();
              if (part.done) {
                observer.finish();
                controller.close();
                reader.releaseLock();
              } else {
                observer.chunk(part.value);
                controller.enqueue(part.value);
              }
            } catch (e) {
              observer.fail(e);
              controller.error(e);
              reader.releaseLock();
            }
          },
          async cancel(reason) {
            observer.fail(reason, true);
            try {
              await reader.cancel(reason);
            } finally {
              reader.releaseLock();
            }
          },
        },
        { highWaterMark: 0 },
      );
      const wrapped = new Response(stream, {
        status: response.status,
        statusText: response.statusText,
        headers: response.headers,
      });
      Object.defineProperties(wrapped, {
        url: { value: response.url },
        redirected: { value: response.redirected },
        type: { value: response.type },
      });
      return wrapped;
    } catch {
      capture.finishPayload(
        handle,
        async () => ({ kind: "absent" }),
        0,
        false,
        "unsupported_http_response",
      );
      return response;
    }
  }) as typeof fetch;
  return wrapped;
}

File: packages/sdk-python/src/hue_sdk/scenes/http.py
"""Scoped HTTP request identities and bounded response tees; never eager-read a body."""

from __future__ import annotations

import json
import re
import zlib
from email.message import Message
from pathlib import PurePosixPath
from urllib.parse import unquote_plus, urlsplit, urlunsplit

from . import _payload
from ._json import canonical_json, clean_json, credential_key, sha256
from .client import Capture, Scenes
from .types import MAX_ARTIFACT_BYTES, SnapshotMissError, SourceFile

REPRESENTATION_HEADERS = frozenset(
    {
        "accept",
        "content-type",
        "range",
        "if-match",
        "if-none-match",
        "if-modified-since",
        "if-unmodified-since",
        "if-range",
    }
)


def is_json_media_type(content_type: str) -> bool:
    media_type = content_type.split(";", 1)[0].strip().lower()
    token = r"[!#$%&'*+\-.^_`|~0-9a-z]+"
    return (
        media_type == "application/json"
        or re.fullmatch(token + "/" + token + r"\+json", media_type) is not None
    )


def parse_http_json(body: bytes):
    """JSON on the wire is UTF-8, with only an optional UTF-8 BOM accepted."""
    return json.loads(body.decode("utf-8-sig", errors="strict"))


def safe_url(url: str) -> str:
    parts = urlsplit(url)
    hostname = parts.hostname
    if hostname is None:
        # Opaque source URIs (e.g. an MCP resource identifier) are retained only without query.
        return urlunsplit((parts.scheme, parts.netloc, parts.path, "", ""))
    hostname = hostname.lower()
    if ":" in hostname:
        hostname = "[" + hostname + "]"
    port = parts.port
    default_port = 443 if parts.scheme.lower() == "https" else 80
    authority = hostname + (f":{port}" if port is not None and port != default_port else "")
    query = "&".join(
        part
        for part in parts.query.split("&")
        if not credential_key(unquote_plus(part.partition("=")[0]))
    )
    return urlunsplit((parts.scheme.lower(), authority, parts.path or "/", query, ""))


def response_headers(headers) -> dict[str, str]:
    return {
        str(key).lower(): str(value)
        for key, value in headers.items()
        if not credential_key(str(key))
    }


def owned_binding(scenes: Scenes, url: str):
    active = scenes._active.get()
    if active is None:
        return None, None
    parsed = urlsplit(safe_url(url))
    origin = urlunsplit((parsed.scheme, parsed.netloc, "", "", ""))
    matches = [
        binding
        for binding in active.selected.values()
        if binding.get("kind") == "http"
        and binding.get("http", {}).get("origin") == origin
        and parsed.path.startswith(binding["http"]["pathPrefix"])
    ]
    if len(matches) > 1:
        if not isinstance(active, Capture):
            raise SnapshotMissError("overlapping_bindings")
        # An explicitly narrower scope owns capture when both are present; replay rejects overlap.
        matches.sort(key=lambda binding: len(binding["http"]["pathPrefix"]), reverse=True)
    return active, matches[0] if matches else None


def http_arguments(binding, method: str, url: str, headers, body: bytes | None):
    normalized_headers = {str(key).lower(): str(value) for key, value in headers.items()}
    selected = REPRESENTATION_HEADERS | {
        name.lower() for name in binding["http"].get("headers", [])
    }
    selected_headers = {
        key: value
        for key, value in normalized_headers.items()
        if key in selected and not credential_key(key)
    }
    if body is None or "multipart/" in normalized_headers.get("content-type", "").lower():
        raise ValueError("Streamed and multipart source request bodies are not portable.")
    if body and is_json_media_type(normalized_headers.get("content-type", "")):
        body = canonical_json(clean_json(parse_http_json(body)))
    return {
        "method": method.upper(),
        "url": safe_url(url),
        "headers": selected_headers,
        "bodySha256": sha256(body),
    }


def decoded_http_body(body: bytes, headers: dict[str, str]) -> bytes:
    """Bounded standard compression codecs; unsupported encodings fail closed."""
    encodings = headers.get("content-encoding", "").lower().split(",")
    for encoding in reversed(encodings):
        encoding = encoding.strip()
        if encoding in {"", "identity"}:
            continue
        if encoding not in {"gzip", "deflate"}:
            raise SnapshotMissError("nonportable")
        options = (31,) if encoding == "gzip" else (zlib.MAX_WBITS, -zlib.MAX_WBITS)
        decoded = None
        for window in options:
            try:
                decompressor = zlib.decompressobj(window)
                decoded = decompressor.decompress(body, MAX_ARTIFACT_BYTES + 1)
                if len(decoded) > MAX_ARTIFACT_BYTES or decompressor.unconsumed_tail:
                    raise SnapshotMissError("nonportable")
                if not decompressor.eof or decompressor.unused_data:
                    raise SnapshotMissError("nonportable")
                break
            except zlib.error:
                continue
        if decoded is None:
            raise SnapshotMissError("nonportable")
        body = decoded
    return body


class ResponseCapture:
    def __init__(
        self,
        call,
        status: int,
        headers,
        url: str,
        status_text: str = "",
        request_method: str = "GET",
    ):
        self.call = call
        self.request_method = request_method
        self.metadata = {
            "kind": "http",
            "status": status,
            "headers": response_headers(headers),
            "url": safe_url(url),
            "statusText": status_text,
        }
        self._body = bytearray()
        self._reserved = 0
        self._omission = None
        self._done = False

    def append(self, chunk: bytes) -> None:
        if self._done or self._omission:
            return
        scenes = self.call.capture.scenes
        if len(self._body) + len(chunk) > MAX_ARTIFACT_BYTES or not scenes._reserve(len(chunk)):
            self._omission = "body_limit"
            scenes._release(self._reserved)
            self._reserved = 0
            self._body.clear()
            return
        self._reserved += len(chunk)
        self._body.extend(chunk)

    def finish(
        self, *, incomplete: bool = False, error: BaseException | None = None, decoded: bool = False
    ) -> None:
        if self._done:
            return
        self._done = True
        scenes = self.call.capture.scenes
        scenes._release(self._reserved)
        self._reserved = 0
        omission = self._omission or ("body_not_consumed" if incomplete else None)
        try:
            if error is not None or omission:
                partial = None
                if self._body and not is_json_media_type(
                    self.metadata["headers"].get("content-type", "")
                ):
                    partial = {**self.metadata, "body": _payload.encode(bytes(self._body))}
                self.call.finish(
                    error=error, payload=partial, omission=omission or "body_interrupted"
                )
            else:
                body = scenes._clean("http_body", bytes(self._body))
                if not isinstance(body, bytes):
                    raise ValueError()
                if body != self._body:
                    self.call.finish(omission="redacted_body")
                    return
                if is_json_media_type(self.metadata["headers"].get("content-type", "")) and (
                    self.request_method != "HEAD" and self.metadata["status"] not in {204, 205, 304}
                ):
                    json_bytes = (
                        body if decoded else decoded_http_body(body, self.metadata["headers"])
                    )
                    value = parse_http_json(json_bytes)
                    if canonical_json(value) != canonical_json(scenes._clean("http_json", value)):
                        self.call.finish(omission="redacted_body")
                        return
                headers = self.metadata["headers"]
                if decoded and body and headers.get("content-encoding"):
                    headers.pop("content-encoding", None)
                    headers.pop("content-length", None)
                headers.pop("transfer-encoding", None)
                payload = _payload.encode(body)
                sources = []
                mime_type = headers.get("content-type", "").split(";")[0].lower().strip()
                document_types = {
                    "application/pdf",
                    "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
                    "application/vnd.openxmlformats-officedocument.presentationml.presentation",
                }
                disposition = headers.get("content-disposition", "")
                if self.request_method != "HEAD" and (
                    mime_type in document_types or disposition.lower().startswith("attachment")
                ):
                    message = Message()
                    message["content-disposition"] = disposition
                    name = (
                        message.get_filename()
                        or PurePosixPath(urlsplit(self.metadata["url"]).path).name
                        or "observed-source"
                    )
                    name = name.replace("\\", "/").rsplit("/", 1)[-1]
                    source_bytes = decoded_http_body(body, headers)
                    sources = [
                        SourceFile(
                            name,
                            source_bytes,
                            mime_type or "application/octet-stream",
                            "tool_source",
                            self.metadata["url"],
                        )
                    ]
                    if source_bytes == body:
                        payload = {
                            "kind": "blob",
                            "ref": _payload.Blob(
                                body, "bytes", sources[0].mime_type, name, "source"
                            ),
                        }
                self.call.finish(payload={**self.metadata, "body": payload}, sources=sources)
        except BaseException:
            self.call.finish(omission="nonportable_body")
        finally:
            self._body.clear()
